QueueImplementation/VM/CustomCollectionViewModel.cs

```diff
@@ -0,0 +1,363 @@
+﻿using Business.Helper;
+using Business.Interfaces;
+using MvvmCross.ViewModels;
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+
+namespace ViewModel
+{
+	/// <summary>
+	/// Represents the view model used for managing the collection.
+	/// </summary>
+	/// <owner>Anton Petrenko</owner>
+	public sealed class CustomCollectionViewModel<T> : MvxViewModel
+	{
+		/// <summary>
+		/// Holds the add value command.
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		private RelayCommand add;
+
+		/// <summary>
+		/// Holds the all numbers.
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		private ObservableCollection<int> allNumbers;
+
+		/// <summary>
+		/// Holds the collection.
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		private ICustomCollection<int> collection;
+
+		/// <summary>
+		/// Holds the configuration creator.
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		private readonly CollectionConfigCreator<int> configCreator = new CollectionConfigCreator<int>();
+
+		/// <summary>
+		/// Holds the current element that will be processed.
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		private int? elementToProcess;
+
+		/// <summary>
+		/// Holds the is choice possible.
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		private bool isChoisePossible = true;
+
+		/// <summary>
+		/// Holds the message to display.
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		private string message;
+
+		/// <summary>
+		/// Holds the remove value command.
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		private RelayCommand remove;
+
+		/// <summary>
+		/// Holds the show value command.
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		private RelayCommand show;
+
+		/// <summary>
+		/// Gets the add command.
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		/// <value>The add command.</value>
+		public RelayCommand Add
+		{
+			get
+			{
+				if (this.add is null)
+					this.add = new RelayCommand(this.AddNumber, this.CanAddNumber);
+
+				return this.add;
+			}
+		}
+
+		/// <summary>
+		/// Adds the number to the collection.
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		/// <param name="obj">The object.</param>
+		private void AddNumber(object obj)
+		{
+			if (this.Collection is null)
+				return;
+
+			int previousCount = this.Collection.Count();
+			this.RestoreCurrentElement();
+			int newValue = RandomNumberGenerator.GetValue();
+			this.Collection.Add(newValue);
+			this.Message = $"The item {newValue} was added.";
+
+			this.RaisePropertyChanged(() => this.AllNumbers);
+
+			if (previousCount == 0)
+				this.RestoreRemoveAndShowCommands();
+		}
+
+		/// <summary>
+		/// Gets the collection of all numbers.
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		/// <value>The collection of all numbers.</value>
+		public IEnumerable<int> AllNumbers
```

Comment:
The property does not meet "generic" requirements. What if we want to store chars immediately?

Reply:
Whole this view model works with the integer type.